### PR TITLE
Add configurable FP8 scaling recipes

### LINKED ISSUE #3967: Expose FP8 Scaling Recipe (Rowwise) Instead of Hardcoded Tensorwise
### ⚠️ Please check that this feature request hasn't been suggested before.

- [x] I searched previous [Ideas in Discussions](https://github.com/axolotl-ai-cloud/axolotl/discussions/categories/ideas) didn't find any similar feature requests.
- [x] I searched previous [Issues](https://github.com/axolotl-ai-cloud/axolotl/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description
https://arxiv.org/abs/2608.27370

fp8: true always trains with tensorwise FP8 scaling. In AxolotlTrainer.additional_accelerator_args (src/axolotl/core/trainers/base.py:682-702) the config is built as a bare Float8LinearConfig(...), which defaults to ScalingGranularity.TENSORWISE. The only knobs plumbed through from YAML are fp8 and fp8_enable_fsdp_float8_all_gather.

torchao (we pin 0.17.0) already ships finer-grained recipes via Float8LinearConfig.from_recipe_name(...):
```
Float8LinearRecipeName.TENSORWISE — current behaviour
Float8LinearRecipeName.ROWWISE — dynamic per-row scaling, CUTLASS rowwise kernel
Float8LinearRecipeName.ROWWISE_WITH_GW_HP — rowwise with the grad-weight GEMM in high precision; more accurate, slightly slower
```
Motivation: tensorwise is the least accurate of the three and the gap widens with token count. The paper runs DeepSeek-V3-style blockwise FP8 from step 0 and reports ~98% BF16-equivalent compute retention with ~25% fewer GPU-hours at matched quality. Their result is specifically that fine-grained scaling is what makes long FP8 runs viable. Rowwise is the closest thing we can offer in-tree.

### ✔️ Solution

Add an fp8_recipe config field and pass it through:

```
fp8: true
fp8_recipe: rowwise        # tensorwise (default) | rowwise | rowwise_with_gw_hp
fp8_enable_fsdp_float8_all_gather: true
torch_compile: true
```

### ❓ Alternatives

_No response_

### 📝 Additional Context

Our FP8 hardware gate already admits consumer Blackwell: compute_supports_fp8() (src/axolotl/cli/config.py:344) is torch.cuda.get_device_capability() >= (9, 0), so an RTX 5090 at sm_120 → (12, 0) passes, and check_fp8_attention_preflight explicitly whitelists sm_120. But docs/mixed_precision.qmd:80 still says FP8 "requires compatible hardware (H100, H200)" — stale, and misleading for exactly the consumer-Blackwell users this helps. Worth fixing in the same PR.

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this feature has not been requested yet.
- [x] I have provided enough information for the maintainers to understand and evaluate this request.

### PR DESCRIPTION
**Pull request by JimmyWang0417**
## Summary

Fixes #3967.

Expose TorchAO's `tensorwise`, `rowwise`, and `rowwise_with_gw_hp` FP8
scaling recipes through the Axolotl config. The default remains `tensorwise`.
FSDP float8 all-gather remains available for tensorwise scaling; incompatible
rowwise combinations are rejected during validation.

The recipe is passed through the CLI metadata, Trainer monkeypatch, and
PatchManager. Documentation now covers compatible Blackwell hardware, and the
tests exercise the pinned TorchAO API and the forwarding path.

## Type of change

- [x] Bug fix
- [x] New feature
- [ ] Breaking change
- [x] Improvement
- [ ] Model update
- [ ] Other:

## How has this been tested?

- Focused CPU tests (TorchAO 0.17): 9 passed.
- The tests cover all three recipes, rowwise/all-gather validation, CLI metadata,
  Trainer patch formatting, and PatchManager forwarding.
- Pinned Ruff check and format check passed.
- Python compilation and `git diff --check` passed.

The full suite requires the repository's GPU/ML dependency set, which was not
available in this environment.

## AI Usage Disclaimer

Yes — Codex assisted with implementation and tests. The changes were checked
against TorchAO 0.17 and the issue requirements.

## Additional Notes

`fp8_enable_fsdp_float8_all_gather: true` remains supported only with the
default `tensorwise` recipe, matching TorchAO 0.17.

## Review update

Following maintainer feedback, the new recipe is now configured as `fp8_config.recipe`, matching the `fsdp_config` nesting pattern. The existing `fp8_enable_fsdp_float8_all_gather` option remains flat for compatibility. Documentation changes were narrowed to the requested hardware guidance and nested examples.

### CODE REVIEW

File: docs/mixed_precision.qmd [hunk 1]
```diff
@@ -82,7 +82,7 @@ FP8 support is experimental and requires compatible hardware (H100, H200) and re
 
 ### What is FP8? {#sec-fp8-overview}
 
-FP8 (8-bit floating point) can provide significant time savings compared to FP16/BF16 while maintaining training stability. Axolotl's implementation uses PyTorch's TorchAO library with "tensorwise" scaling strategy.
+FP8 (8-bit floating point) can provide significant time savings compared to FP16/BF16 while maintaining training stability. Axolotl's implementation uses PyTorch's TorchAO library and supports tensorwise and rowwise scaling recipes.
```
[the remark this answers is not in the document]
Comment (JimmyWang0417) [reply docs/mixed_precision.qmd]: Thanks for catching this. I updated the FP8 hardware guidance to include compatible Blackwell devices and added the TorchAO 0.17 API and PatchManager forwarding tests. The revised checks pass.

File: src/axolotl/utils/schemas/validation.py [hunk 2]
```diff
@@ -425,6 +425,17 @@ def check_fft_possible_bad_config(self):
     @model_validator(mode="before")
     @classmethod
     def check_fp8_config(cls, data):
+        fp8_config = data.get("fp8_config") or {}
+        fp8_recipe = (
+            fp8_config.get("recipe", "tensorwise")
+            if isinstance(fp8_config, dict)
+            else "tensorwise"
```
Comment (ved1beta) [inline src/axolotl/utils/schemas/validation.py]: fp8_config is validated independently of fp8. Setting fp8_config: {recipe: rowwise} without fp8: true is silently ignored

**Review by ved1beta — CHANGES_REQUESTED** [review]

File: src/axolotl/utils/schemas/validation.py (hunk 2 above)
Comment (JimmyWang0417) [reply src/axolotl/utils/schemas/validation.py]: Fixed in 6fecf22. A non-empty `fp8_config` now requires `fp8: true` and raises a clear `ValidationError` instead of being silently ignored. I added a regression test and updated the rowwise/FSDP fixture. An isolated validation harness passes all 3 checks. The full repository pytest fixture was blocked by unavailable HF downloads, not by a failed assertion. Environment: RTX 5090 32 GB, driver 595.71.05, CUDA 13.0, PyTorch 2.12.1+cu130, TorchAO 0.17.0, Transformers 5.16.1, Accelerate 1.13.0.

**Review by ved1beta — APPROVED** [review]
great work !
